### PR TITLE
fix(#792): stop the root-implicit path silently ignoring four documented knobs

### PR DESCRIPTION
**Pull request by yingjerkao**
> 🤖 **AI-generated PR** — written by Claude Code, posted by @yingjerkao.

Four documented settings were accepted by `optimize_gs_ad_root_implicit` and then dropped, so the caller got a materially different optimizer with no signal.

**They do not all get the same treatment**, and that distinction is the substance of this PR:

| knob | default | treatment | why |
|---|---|---|---|
| `gs_optimizer=cg` | opt-in | **raise** | runs plain gradient descent under the CG name |
| `gs_c4v=True` | opt-in | **raise** | nothing projects into the C4v basis |
| `gs_line_search` | **auto-ON** | **warn** | refusing it rejects the path's own default config |
| `return_history` | opt-in | **implement** | documented API; the loop already had the data |

## The trap: `gs_line_search` cannot be refused

The issue groups all four and suggests rejecting them at the dispatcher. That would have been wrong for this one.

`gs_optimizer` defaults to `lbfgs`, and `gs_line_search=None` means *auto-on* for lbfgs/cg (`_use_line_search`, `ipeps_optimize.py:476-480`). A blanket rejection therefore raises on the **default** root-implicit configuration — and on the documented Path 5 example.

So it warns, following the rule this file already established for `gs_metric_precond` (*"Default-on knob, so this warns rather than raising"*). That rule is now written down at the validation gate: **refuse what was explicitly requested, warn about what arrived by default.**

The warning is accurate about what is actually happening: `_build_optimizer` returns a plain `optax.chain(scale_by_lbfgs, clip_by_global_norm, scale(-1))`, and `value_fn=` is consumed only by `optax.lbfgs(linesearch=...)`. Passing it to that chain does nothing. Steps are unconditional quasi-Newton steps and an energy-increasing one can be accepted — the pre-existing in-code comment read as though a search were running.

## `return_history` implemented, not refused

Energies, per-step times and the convergence flag were all computed and discarded; only recording was missing. Same dict shape *and the same shape invariant* as the 1-site tensor path in `ipeps_optimize`:

```
len(step_times) == max(num_steps - 1, 0)
```

Step 0 is charged to `jit_compile_time` and excluded from `step_times`. This path has no top-level `jax.jit`, but its root solve and adjoint are jitted internally, so the first evaluation is still the compile-dominated one.

The `gs_num_steps == 0` early return is covered too; otherwise it would return a 3-tuple while the main path returned 4.

This is the gap that forced a two-point compile/steady-state fit in the AD path benchmark, so it has an immediate consumer.

## Verification

Mutation-verified, each killing exactly its own tests:

| mutation | kills |
|---|---|
| drop cg rejection | cg test + names-every-knob test |
| drop c4v rejection | c4v test + names-every-knob test |
| drop the warning | the warns test |
| **warn unconditionally** | the stays-silent test |
| `if first_step` → `if False` | the `step_times`/`jit_compile_time` shape test |
| `converged_flag = True` | the masked-gradient test |

The last two are complementary on purpose: together they prove the warning is **conditional in both directions**, not merely present. A warning that always fires is noise, and `gs_line_search=False` — what the documented example uses — must stay quiet.

There is also an explicit test that the guard does **not** reject the default configuration, which is the regression this design is most likely to introduce.

Tests: **39 passed** across the new file, `test_root_implicit_wiring.py` and `test_ipeps_ad_history.py` (the contract this path now conforms to).

## Review round 2 (Codex, both confirmed)

**P1 — cold first solve counted as a warm step.** Confirmed, and worse than reported: it broke the invariant `tests/test_ipeps_ad_history.py:62` asserts for the other AD paths, so "same dict shape" was not true. The named consumer in this very PR — the benchmark's two-point compile/steady-state fit — would have been handed a cold sample as warm and told compilation was free. Fixed; the `jit_compile_time = 0.0` rationale above was wrong and has been corrected.

**P2 — convergence reported after masking a failed gradient.** Confirmed. A fully masked gradient has L2 norm exactly `0.0`, so `_converged_outer` fires on a step this loop already warns is "a no-op and apparent convergence here is not convergence", and whose run-level warning declares the whole run "not ... a converged optimization". The flag now agrees with the warnings.

Scope: only the **flag** is fixed here. The `break` on fake convergence is pre-existing, affects every root-implicit caller rather than just `return_history` ones, and shifts returned energies and step counts — filed as #812 with a suggested fix modelled on the existing `prev_energy = float("inf")` guard in `ipeps_optimize.py`.

## CI

GitHub Actions has been in [major outage](https://www.githubstatus.com) since 15:22 UTC and still is. No workflow run has been created **repo-wide** since 19:20 UTC, so `43558ae` and `3ea6c1b` both have zero check-runs. This is infrastructural — nothing on this branch. Checks should populate on their own once Actions recovers; no further push is needed.

Verified locally in the meantime (`JAX_PLATFORMS=cpu`): **39 passed** across `test_root_implicit_ignored_knobs.py`, `test_root_implicit_wiring.py` and `test_ipeps_ad_history.py`; `ruff check` and `ruff format` clean. The `core` subset of the new file — what the required gate actually runs — is 8 tests in ~5s wall.

### CODE REVIEW

File: src/tenax/algorithms/ipeps_optimize_root_implicit.py [hunk 1]
```diff
@@ -308,6 +389,8 @@ def _final_env(params):
             )
         grads = jnp.where(jnp.isfinite(grads), grads, 0.0)
         E = float(jnp.real(energy_val))
+        history_energies.append(E)
+        history_step_times.append(_time.perf_counter() - _step_t0)
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/ipeps_optimize_root_implicit.py]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Confirmed, and the impact is larger than stated. This did not just mislabel a number — it broke a **structural invariant** the other AD paths are held to, `tests/test_ipeps_ad_history.py:62`:

```python
# step_times excludes step 0 (which lands in jit_compile_time), so
# len(step_times) == len(energies) - 1 when num_steps > 1, or 0 if num_steps == 1.
assert len(history["step_times"]) == max(history["num_steps"] - 1, 0)
```

Appending every step made `len(step_times) == num_steps`, so the PR body claim of "same dict shape as the 1-site tensor path" was false. And the consumer this PR names as its motivation — the AD-path benchmark two-point compile/steady-state fit — is exactly the one that would have been corrupted: a cold sample presented as warm, plus `jit_compile_time = 0.0` asserting compilation was free.

You are also right on the mechanism. No top-level `jax.jit` on this path does not mean no compilation: the root solve and its adjoint are jitted internally, so step 0 pays their lazy compile.

Fixed in 3ea6c1b by mirroring the `_first_step` diversion in `ipeps_optimize.py:2006-2013`. Mutation-verified: `if first_step:` → `if False:` fails the shape assertion `2 == 1` and kills that test alone.

File: src/tenax/algorithms/ipeps_optimize_root_implicit.py [hunk 2]
```diff
@@ -342,6 +425,7 @@ def _final_env(params):
                     grad_norm_tol=config.gs_grad_norm_tol,
                     criterion=config.gs_conv_criterion,
                 )
+            converged_flag = True
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/ipeps_optimize_root_implicit.py]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Confirmed. A fully masked gradient has L2 norm **exactly** `0.0`, which is below any `gs_grad_norm_tol`, so `_converged_outer` fires on a step this same function has already warned is

> every entry was non-finite, so the step is a no-op and **apparent convergence here is not convergence**

Worth adding that `gs_conv_criterion='dE'` — the current default — has the same hole one step later: a fully masked step is a no-op, so the next step re-evaluates identical params and sees `delta_energy == 0.0`.

The decisive argument for gating run-wide rather than per-step is that the loop's existing end-of-run warning already declares the **whole run** non-converged whenever any step was contaminated ("The reported energy is from the best state actually reached, not from a converged optimization"). `converged=True` alongside that warning would be a direct self-contradiction, so the flag is now `nonfinite_grad_steps == 0`.

Mutation-verified: `converged_flag = True` fails `assert True is False` and kills that test alone.

**Scope note.** Only the flag is fixed here — I deliberately did *not* change the `break`. That is pre-existing behaviour affecting every root-implicit caller, not just `return_history` ones, and suppressing it shifts returned energies and step counts. Filed as #812, with a suggested fix modelled on the `prev_energy = float("inf")` guard that `ipeps_optimize.py` already uses for the analogous post-rollback false-convergence case.
